### PR TITLE
Vram measurement

### PR DESCRIPTION
**Pull request by andimarafioti**
Added a small script to measure the vram and run it on an H100.

### CODE REVIEW

**Comment by andimarafioti** [comment]
This addresses: https://github.com/huggingface/nanoVLM/issues/25

File: measure_vram.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    parser = argparse.ArgumentParser(description="Measure VRAM usage for a VisionLanguageModel at different batch sizes.")
+    
+    # Model and Config args
+    parser.add_argument('--vlm_checkpoint_path', type=str, default=None, help='Path to the VLM checkpoint to load. If None, initializes a new model based on VLMConfig.')
+    parser.add_argument('--compile', action='store_true', help='Compile the model with torch.compile.')
+
+    # Measurement control args
+    parser.add_argument('--batch_sizes', type=str, default="1 2 4", help='Space-separated list of batch sizes to test (e.g., "1 2 4 8").')
+    parser.add_argument('--num_iterations', type=int, default=2, help='Number of forward/backward passes per batch size for VRAM measurement.')
+    parser.add_argument('--num_workers', type=int, default=0, help='Number of workers for DataLoader. 0 is often best for precise VRAM measurement to avoid overhead.')
```
Comment (ariG23498) [inline measure_vram.py:183]: If this is the best practise as suggested in the description, how about we make it default and not give users an argument to play with? Just thinking out loud here.

File: measure_vram.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    train_cfg_defaults = config.TrainConfig() # Used for default dataset path/name if not provided by CLI
+
+    # Update VLMConfig from CLI if needed (e.g. if more vlm_cfg options were added to parser)
+    # For now, only vlm_checkpoint_path directly influences model loading path.
+    # Other vlm_cfg settings (vit_img_size, lm_tokenizer etc.) are taken from models.config.VLMConfig()
+
+    print("--- VLM Config (from models.config) ---")
+    print(vlm_cfg) # Show base config
+    if args.vlm_checkpoint_path:
+        print(f"CLI override: vlm_checkpoint_path = {args.vlm_checkpoint_path}")
```
Comment (ariG23498) [inline measure_vram.py:197]: I don't think this is a good idea. We should rather ask the user to modify the configuration themselves. I have always found out code which overwrite a config turn out to be problematic later.

File: measure_vram.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    # --- Model Initialization ---
+    if args.vlm_checkpoint_path:
+        print(f"Loading model from specified checkpoint: {args.vlm_checkpoint_path}")
+        model = VisionLanguageModel.from_pretrained(args.vlm_checkpoint_path)
+    else:
+        print(f"Initializing a new model (no specific checkpoint path provided via CLI).")
+        print(f"Using VLMConfig defaults: load_backbone_weights={vlm_cfg.vlm_load_backbone_weights}")
+        model = VisionLanguageModel(vlm_cfg, load_backbone=vlm_cfg.vlm_load_backbone_weights)
+    
+    model.to(device)
```
Comment (ariG23498) [inline measure_vram.py:35]: Do you think we should also print out how much VRAM is needed for the model? This would be a very small addition to code, but later on if users wanted to play around with the architecture this would be helpful to know.

File: measure_vram.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    if args.compile:
+        print("Compiling the model with torch.compile...")
+        model = torch.compile(model)
+        print("Model compiled.")
```
Comment (ariG23498) [inline measure_vram.py:39]: Should we not compile the model first and then send it to device?

File: measure_vram.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    dataset_path = train_cfg_defaults.train_dataset_path
+    # train_cfg_defaults.train_dataset_name is a list, use the first if not specified
+    dataset_name = train_cfg_defaults.train_dataset_name[0] if train_cfg_defaults.train_dataset_name else None
+
+    batch_sizes_to_test = [int(bs) for bs in args.batch_sizes.split()]
+    if not batch_sizes_to_test:
+        print("Error: No batch sizes provided or parsed correctly.")
+        return
+    
+    num_iterations_for_vram = args.num_iterations
```
Comment (ariG23498) [inline measure_vram.py:56]: Just curious, why do we need more than one iterations? If this is due to the fact that the first iteration is considered to be a warmup for some kernels, then I think adding a condition of `>1` is necessary.

File: measure_vram.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+        try:
+            for i, batch in enumerate(current_loader):
+                if i >= num_iterations_for_vram:
+                    break
+                
+                images = batch["image"].to(device)
+                input_ids = batch["input_ids"].to(device)
+                labels = batch["labels"].to(device)
+                attention_mask = batch["attention_mask"].to(device)
```
Comment (ariG23498) [inline measure_vram.py:130]: I think we should also have a VRAM print for the inputs. This also helps with fine grained details. Please feel free to ignore this if this is not important to the end user.

File: measure_vram.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+                optimizer.zero_grad(set_to_none=True)
+
+                with torch.autocast(device_type='cuda', dtype=torch.bfloat16 if torch.cuda.is_bf16_supported() else torch.float16):
+                    _, loss = model(input_ids, images, attention_mask=attention_mask, targets=labels)
```
Comment (ariG23498) [inline measure_vram.py:135]: I am sceptical about this. If we do autocast, we should be printing that out as information. WDYT?

**Review by ariG23498 — CHANGES_REQUESTED** [review]
I love the fact that this is very straight-forward and people will find great value with this script.

I also think we should generate a file rather than printing the information to terminal. What this would do is for users to have artefacts which they can share over and also compare.

I also think we should have a similar setup for `inference`. Looking forward to what you think.

File: measure_vram.py (hunk 1 above)
Comment (ariG23498) [reply measure_vram.py:183]: I see that I have not tagged the lines correctly, I wanted to point to only `num_workers` and not the other arguments.

File: measure_vram.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    for bs, vram_usage in results.items():
+        print(f"Batch Size {bs}: {vram_usage}")
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Measure VRAM usage for a VisionLanguageModel at different batch sizes.")
+    
+    # Model and Config args
+    parser.add_argument('--vlm_checkpoint_path', type=str, default=None, help='Path to the VLM checkpoint to load. If None, initializes a new model based on VLMConfig.')
+    parser.add_argument('--compile', action='store_true', help='Compile the model with torch.compile.')
```
[the remark this answers is not in the document]
Comment (andimarafioti) [reply measure_vram.py:178]: We are, I just gave the option to only pass the checkpoint path as I thought it might be easier than creating a whole config.

File: measure_vram.py (hunk 1 above)
Comment (andimarafioti) [reply measure_vram.py:183]: Sure, I removed the argument

File: measure_vram.py (hunk 2 above)
Comment (andimarafioti) [reply measure_vram.py:197]: Ok, I removed the vlm_checkpoint path since you complained twice xD

File: measure_vram.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (andimarafioti) [reply measure_vram.py:178]: nvm, ended up removing it :)

File: measure_vram.py (hunk 4 above)
Comment (andimarafioti) [reply measure_vram.py:39]: I'm not sure, but I moved the compilation to before moving it to the device. The previous version was working tho

File: measure_vram.py (hunk 5 above)
Comment (andimarafioti) [reply measure_vram.py:56]: If the dataloader is not very well written, some batches will be longer than others. And longer batches take more RAM.

File: measure_vram.py (hunk 6 above)
Comment (andimarafioti) [reply measure_vram.py:130]: Too detailed right now, let's see what people say

File: measure_vram.py (hunk 5 above)
Comment (ariG23498) [reply measure_vram.py:59]: Interesting point. So if I understand this correctly, more number of iterations mean we take a better average?

Hearing this I would say we need to fix the inputs. We cannot have this dynamically set, which would lead to different VRAM numbers for different datasets. What do you think?

Comment (andimarafioti) [reply measure_vram.py:59]: Future work :)

**Review by ariG23498 — APPROVED** [review]
LGTM!

Thanks for iterating on this.
